src/databricks/labs/ucx/framework/owners.py

```diff
@@ -0,0 +1,204 @@
+        return (user for user in all_users if user.active and user.user_name and self._has_role(user, "account_admin"))
+
+
+class AdministratorLocator:
+    """Locate a workspace administrator, if possible.
+
+    This will first try to find an active workspace administrator. If there are multiple, the first (alphabetically
+    sorted by user-name) will be used. If no active workspace administrators can be found then an account administrator
+    is sought, again returning the first alphabetically by user-name if more than one is found.
+    """
+
+    def __init__(
+        self,
+        ws: WorkspaceClient,
+        *,
+        finders: Sequence[Callable[[WorkspaceClient], AdministratorFinder]] = (
+            WorkspaceAdministratorFinder,
+            AccountAdministratorFinder,
+        ),
+    ) -> None:
+        """
+        Initialize the instance, which will try to locate administrators using the workspace for the supplied client.
+
+        Args:
+            ws (WorkspaceClient): the client for workspace in which to locate admin users.
+            finders: a sequence of factories that will be instantiated on demand to locate admin users.
+        """
+        self._ws = ws
+        self._finders = finders
+
+    @cached_property
+    def _workspace_id(self) -> int:
+        # Makes a REST call, so we cache it.
+        return self._ws.get_workspace_id()
+
+    @cached_property
+    def _found_admin(self) -> str | None:
+        # Lazily instantiate and query the finders in an attempt to locate an admin user.
+        finders = (finder(self._ws) for finder in self._finders)
+        # If a finder returns multiple admin users, use the first (alphabetically by user-name).
+        first_user = functools.partial(min, default=None, key=lambda user: user.user_name)
+        found_admin_users: Iterable[User | None] = (first_user(finder.find_admin_users()) for finder in finders)
+        return next((user.user_name for user in found_admin_users if user), None)
+
+    @property
+    def workspace_administrator(self) -> str:
+        """The user-name of an admin user for the workspace.
+
+        Raises:
+              RuntimeError if an admin user cannot be found in the current workspace.
+        """
+        found_admin = self._found_admin
+        if found_admin is None:
+            msg = f"No active workspace or account administrator can be found for workspace: {self._workspace_id}"
+            raise RuntimeError(msg)
+        return found_admin
+
+
+class Ownership(ABC, Generic[Record]):
+    """Determine an owner for a given type of object."""
+
+    def __init__(self, administrator_locator: AdministratorLocator) -> None:
+        self._administrator_locator = administrator_locator
+
+    @final
+    @property
+    def administrator_locator(self):
+        return self._administrator_locator
+
+    @final
+    def owner_of(self, record: Record) -> str:
```

Comment:
i am struggling to find where this method is called in production code. shouldn't it be called from history recorder?

Reply:
Absolutely: this method is not being accessed yet. It will be called from the history recorder, but this PR blocks the history recorder (#2743). The history recorder will be updated to use this interface.